internal/pkg/manifest/validate_integration_test.go

```diff
@@ -0,0 +1,87 @@
+// +build integration localintegration
+
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package manifest_test
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/aws/copilot-cli/internal/pkg/manifest"
+	"github.com/stretchr/testify/require"
+)
+
+var basicKinds = []reflect.Kind{
+	reflect.Bool,
+	reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+	reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
+	reflect.Float32, reflect.Float64,
+	reflect.Complex64, reflect.Complex128,
+	reflect.Array, reflect.String, reflect.Slice, reflect.Map,
+}
+
+type validator interface {
+	Validate() error
+}
+
+// Test_ValidateAudit ensures that every manifest struct implements "Validate()" method.
+func Test_ValidateAudit(t *testing.T) {
+	testCases := map[string]struct {
+		mft manifest.WorkloadManifest
+	}{
+		"backend service": {
+			mft: manifest.BackendService{},
+		},
+		"load balanced web service": {
+			mft: manifest.LoadBalancedWebService{},
+		},
+		"request-driven web service": {
+			mft: manifest.RequestDrivenWebService{},
+		},
+		"schedule job": {
+			mft: manifest.ScheduledJob{},
+		},
+		"worker service": {
+			mft: manifest.WorkerService{},
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			err := isValid(reflect.ValueOf(tc.mft))
+			require.NoError(t, err)
+		})
+	}
+}
+
+func isValid(v reflect.Value) error {
+	typ := v.Type()
+	if typ.Kind() == reflect.Ptr {
+		typ = typ.Elem()
+	}
+	for _, k := range basicKinds {
+		if typ.Kind() == k {
+			return nil
+		}
+	}
+	var val validator
+	validatorType := reflect.TypeOf(&val).Elem()
+	if !typ.Implements(validatorType) {
+		return fmt.Errorf(`%v does not implement "Validate()"`, typ)
+	}
+	if typ.Kind() != reflect.Struct {
+		return nil
+	}
+	for i := 0; i < typ.NumField(); i++ {
+		// Skip if it is not a manifest yaml field.
+		if typ.Field(i).Tag.Get("yaml") == "" {
```

Comment:
That's true. I guess I'd rather revert this change @efekarakus

Reply:
😭 I'm just not a fan because it feels so fragile otherwise. 
As a requirement, it makes sense to me to read "any field used during unmarshalling YAML should be validated", but now we lose that nice definition.

Isn't it good enough to have the `Validate` method on `HealthCheckArgsOrString` ? since that already will call internally `Validate` on any of its exclusive subfields. I'd rather have us keep `typ.Field(i).Tag.Get("yaml") == ""`